File: day2/solution.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <cstring>
long long firstInvalidIdSum = 0;
long long secondInvalidIdSum = 0;

void checkForInvalidId(const long id)
{
    std::string idStr = std::to_string(id);
    size_t len = idStr.length();
    for (int dividend = 2; dividend <= (len / 2) + 1; ++dividend)
    {
        if (len % dividend == 0)
        {
            std::string part = idStr.substr(0, len / dividend);
            if(part[0] == '0')
            {
                continue;
            }
            std::string constructed;
            for (int i = 0; i < dividend; ++i)
            {
                constructed += part;
            }
            if (constructed == idStr)
            {
                if (dividend == 2)
                {
                    firstInvalidIdSum += id;
                }
                secondInvalidIdSum += id;
                return;
            }
        }
    }
    // Special check for single digit repeated len times
    if (len > 1 && idStr[0] != '0')
    {
        bool allSame = true;
        for (size_t i = 1; i < len; ++i)
        {
            if (idStr[i] != idStr[0])
            {
                allSame = false;
                break;
            }
        }
        if (allSame)
        {
            secondInvalidIdSum += id;
        }
    }
}
// Find invalid id in the ranges (including start and end) and add them up. Invalid id ranges are  any ID which is made only of some sequence of digits repeated twice.
// Leading zeros do not count
// ID is invalid if it is made only of some sequence of digits repeated at least twice. So, 12341234 (1234 two times),
// 123123123 (123 three times), 1212121212 (12 five times), and 1111111 (1 seven times) are all invalid IDs.
void solve(const std::vector<std::pair<long, long>> &idRanges)
{
    for (const auto &range : idRanges)
    {
        for (long id = range.first; id <= range.second; ++id)
        {
            checkForInvalidId(id);
        }
    }
}

void read(const std::string &filename, std::vector<std::pair<long, long>> &idRanges)
{
    // 1. Read all lines into a vector
    std::ifstream infile(filename);

    std::string line;
    // Read string from file, file has just one string made up of ranges, ranges are number-number, separated by commas
    const char delim = ',';
    while (std::getline(infile, line, delim))
    {
        size_t dashPos = line.find('-');
        if (dashPos != std::string::npos)
        {
            std::cout << "Processing line: " << line << "\n";
            long start = std::stol(line.substr(0, dashPos));
            long end = std::stol(line.substr(dashPos + 1));
            idRanges.emplace_back(start, end);
        }
    }
}

int main(int argc, char **argv)
{
    std::vector<std::pair<long, long>> idRanges;
    // Use a function call to cleanly execute the logic
    read(argv[1], idRanges);
    // Debugging print to output read ranges
    for (const auto &range : idRanges)
    {
        std::cout << "Read range: " << range.first << "-" << range.second << "\n";
    }
    solve(idRanges);
    std::cout << "Invalid ranges : " << firstInvalidIdSum << "\n";
    std::cout << "Invalid ranges : " << secondInvalidIdSum << "\n";
    return 0;
}

File: day2/solution.py
import sys

# Global variables to track sums
first_invalid_id_sum = 0
second_invalid_id_sum = 0

def check_for_invalid_id(id_val: int):
    """
    Checks if an ID is 'invalid' based on repeating patterns.
    - first_invalid_id_sum: IDs repeating exactly twice (e.g., 12341234)
    - second_invalid_id_sum: IDs repeating 2 or more times (e.g., 111, 121212)
    """
    global first_invalid_id_sum, second_invalid_id_sum
    
    id_str = str(id_val)
    length = len(id_str)
    
    # Check for repeating patterns by dividing the length
    # dividend is the number of times the substring repeats
    for dividend in range(2, (length // 2) + 2):
        if length % dividend == 0:
            part_len = length // dividend
            part = id_str[:part_len]
            
            # Leading zeros do not count as valid parts
            if part[0] == '0':
                continue
            
            # Python string multiplication: '123' * 3 = '123123123'
            if part * dividend == id_str:
                if dividend == 2:
                    first_invalid_id_sum += id_val
                
                second_invalid_id_sum += id_val
                return

    # Special check for single digit repeated length times (e.g., 1111)
    # This covers cases where dividend > (length/2)+1 (like 111 where dividend is 3)
    if length > 1 and id_str[0] != '0':
        if all(char == id_str[0] for char in id_str):
            second_invalid_id_sum += id_val

def solve(id_ranges):
    for start, end in id_ranges:
        for id_val in range(start, end + 1):
            check_for_invalid_id(id_val)

def read_ranges(filename):
    id_ranges = []
    try:
        with open(filename, 'r') as f:
            # File contains ranges separated by commas, e.g., "100-200,500-600"
            content = f.read()
            parts = content.split(',')
            for part in parts:
                part = part.strip()
                if '-' in part:
                    print(f"Processing line: {part}")
                    start_str, end_str = part.split('-')
                    id_ranges.append((int(start_str), int(end_str)))
    except FileNotFoundError:
        print(f"Error: File {filename} not found.")
    return id_ranges

if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Usage: python script.py <filename>")
    else:
        ranges = read_ranges(sys.argv[1])
        for r in ranges:
            print(f"Read range: {r[0]}-{r[1]}")
            
        solve(ranges)
        
        print(f"Invalid ranges (Exact twice) : {first_invalid_id_sum}")
        print(f"Invalid ranges (Total invalid) : {second_invalid_id_sum}")